pkg/pillar/types/global.go

```diff
@@ -114,6 +124,185 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
+	boolDefault     bool
+	triStateDefault TriState
+}
+
+type ConfigItemSpecMap struct {
+	globalSettings map[string]ConfigItemSpec
+	agentSettings  map[string]ConfigItemSpec
+}
+
+func (configMap *ConfigItemSpecMap) AddIntItem(key string, defaultInt uint32, min uint32, max uint32, agentSetting bool, globalSetting bool) {
+	log.Debugf("Adding int item %s", key)
+	if defaultInt < min || defaultInt > max {
+		log.Fatalf("Adding int item %s failed. Value does not meet given min/max criteria", key)
+	}
+	configItem := ConfigItemSpec{
+		ItemType:   CONFIGITEM_TYPE_INT,
+		Key:        key,
+		intDefault: defaultInt,
+		intMin:     min,
+		intMax:     max,
+	}
+	if agentSetting {
+		configMap.agentSettings[key] = configItem
+	}
+	if globalSetting {
+		configMap.globalSettings[key] = configItem
+	}
+	log.Debugf("Added int item %s", key)
+}
+
+func (configMap *ConfigItemSpecMap) AddBoolItem(key string, defaultBool bool, agentSetting bool, globalSetting bool) {
+	log.Debugf("Adding bool item %s", key)
+	configItem := ConfigItemSpec{
+		ItemType:    CONFIGITEM_TYPE_BOOL,
+		Key:         key,
+		boolDefault: defaultBool,
+	}
+	if agentSetting {
+		configMap.agentSettings[key] = configItem
+	}
+	if globalSetting {
+		configMap.globalSettings[key] = configItem
+	}
+	log.Debugf("Added bool item %s", key)
+}
+
+func (configMap *ConfigItemSpecMap) AddStringItem(key string, defaultString string, agentSetting bool, globalSetting bool) {
+	log.Debugf("Adding string item %s", key)
+	configItem := ConfigItemSpec{
+		ItemType:      CONFIGITEM_TYPE_STRING,
+		Key:           key,
+		stringDefault: defaultString,
+	}
+	if agentSetting {
+		configMap.agentSettings[key] = configItem
+	}
+	if globalSetting {
+		configMap.globalSettings[key] = configItem
+	}
+	log.Debugf("Added string item %s", key)
+}
+
+func (configMap *ConfigItemSpecMap) AddTriStateItem(key string, defaultTriState TriState, agentSetting bool, globalSetting bool) {
+	log.Debugf("Adding tristate item %s", key)
+	configItem := ConfigItemSpec{
+		Key:             key,
+		ItemType:        CONFIGITEM_TYPE_TRISTATE,
+		triStateDefault: defaultTriState,
+	}
+	if agentSetting {
+		configMap.agentSettings[key] = configItem
+	}
+	if globalSetting {
+		configMap.globalSettings[key] = configItem
+	}
+	log.Debugf("Added tristate item %s", key)
+}
+
+type GlobalConfigItem struct {
+	Key      string
+	ItemType ConfigItemType
+
+	intValue	    uint32
+	stringValue		string
+	boolValue	    bool
+	triStateValue	TriState
+}
+
+type GlobalConfigItemMap struct {
+	globalSettings map[string]GlobalConfigItem
+	agentSettings  map[string]GlobalConfigItem
+}
+
+func (specMap *ConfigItemSpecMap) parseItem(configMap *GlobalConfigItemMap, item config.ConfigItem) error {
+	key := item.Key
+	configItem := GlobalConfigItem {
+		Key:	key,
+	}
+	configSpecMap := specMap.globalSettings
+	if _, ok := specMap.globalSettings[key]; ok {
+		configSpecMap = specMap.globalSettings
+	} else if _, ok := specMap.agentSettings[key]; ok {
```

Comment:
For agent settings - you need to do some processing before the setting:

if key.startsWith("agent.") {
    //Match the following format with a regular expression - agent.<agentname>.<setting>
    use agentname and setting to get the configSpec   
    return 
     < you might even spin this part into a separate routine of this gets big enough >
} else {
    // Check Global settings
    If not found in global settings - Go to that default case..
}